sklearn/metrics/_classification.py

```diff
@@ -142,6 +146,30 @@ def _weighted_sum(sample_score, sample_weight, normalize=False):
         return sample_score.sum()
 
 
+def _nan_average(scores, weights):
+    """
+    Wrapper for np.average, with np.nan values being ignored from the average
+    If no weights are passed, then just same as np.nanmean. Otherwise, do similar
+    to it but with weights as in np.average
+    """
+
+    if weights is None:
+        return np.nanmean(scores)
+
+    mask = np.isnan(scores)
+    if mask.all():
+        return np.nan
+
+    weights = np.asarray(weights)
+    scores, weights = scores[~mask], weights[~mask]
```

Comment:
We can early return when there are no nans:

```suggestion
    if not mask.any():
        return np.average(scores, weights=weights)

    scores, weights = scores[~mask], weights[~mask]
```

Here is a quick benchmark:

```python
import numpy as np

n_samples = 100_000
rng = np.random.default_rng(0)
scores = rng.standard_normal(size=n_samples)
weights = rng.poisson(size=n_samples)

%%timeit
_ = _nan_average(scores, weights=weights)
# 400 µs ± 19.4 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

%%timeit
_ = _nan_average_with_early_return(scores, weights=weights)
# 138 µs ± 4.64 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

%%timeit
_ = np.average(scores, weights=weights)
# 103 µs ± 4.56 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

There is still a runtime regression when compared with `main`. This type of regression is appears when working with small `n_samples`:

```python
import numpy as np
from sklearn.metrics import precision_recall_fscore_support

rng = np.random.default_rng(42)

n_samples = 100
y_true = rng.integers(low=0, high=100, size=n_samples)
y_pred = rng.integers(low=0, high=100, size=n_samples)
weights = rng.poisson(size=n_samples)

## main
%%timeit
_ = precision_recall_fscore_support(y_true, y_pred, beta=0.5, average="micro")
# 262 µs ± 8.92 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

## PR with above suggestion
%%timeit
_ = precision_recall_fscore_support(y_true, y_pred, beta=0.5, average="micro")
# 300 µs ± 10.6 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```

Overall, I think this is acceptable, but I would want to see other opinions.